### PR TITLE
arch: common: multilevel irq: fix issues where the intc device is NULL

### PR DESCRIPTION
**Pull request by ycsin**
Fixed a bug in `multilevel_irq.c` & a bug in the index test for multi-plic build, refactored and updated the test quite a bit.

See each commit message for details.

Tested with `renode_riscv32`, see #65500:

```
(.venv) ycsin@LAPTOP-ROG:~/zephyrproject$ west build -b renode_riscv32 -p always -t run  zephyr/tests/kernel/interrupt/ && cp build/compile_commands.json .
-- west build: making build dir /home/ycsin/zephyrproject/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/ycsin/zephyrproject/zephyr/tests/kernel/interrupt
-- CMake version: 3.27.7
-- Found Python3: /home/ycsin/zephyrproject/.venv/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/ycsin/.cache/zephyr
-- Zephyr version: 3.5.99 (/home/ycsin/zephyrproject/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: renode_riscv32
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.3 (/home/ycsin/zephyr-sdk-0.16.3)
-- Found toolchain: zephyr 0.16.3 (/home/ycsin/zephyr-sdk-0.16.3)
-- Found Dtc: /home/ycsin/zephyr-sdk-0.16.3/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /home/ycsin/zephyrproject/zephyr/boards/riscv/renode_riscv32/renode_riscv32.dts
-- Generated zephyr.dts: /home/ycsin/zephyrproject/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /home/ycsin/zephyrproject/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /home/ycsin/zephyrproject/build/zephyr/dts.cmake
Parsing /home/ycsin/zephyrproject/zephyr/Kconfig
Loaded configuration '/home/ycsin/zephyrproject/zephyr/boards/riscv/renode_riscv32/renode_riscv32_defconfig'
Merged configuration '/home/ycsin/zephyrproject/zephyr/tests/kernel/interrupt/prj.conf'
Configuration saved to '/home/ycsin/zephyrproject/build/zephyr/.config'
Kconfig header saved to '/home/ycsin/zephyrproject/build/zephyr/include/generated/autoconf.h'
-- Found GnuLd: /home/ycsin/zephyr-sdk-0.16.3/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.2.0/../../../../riscv64-zephyr-elf/bin/ld.bfd (found version "2.38") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/ycsin/zephyr-sdk-0.16.3/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc
-- Configuring done (15.3s)
-- Generating done (0.1s)
-- Build files have been written to: /home/ycsin/zephyrproject/build
-- west build: running target run
[1/111] Preparing syscall dependency handling

[2/111] Generating include/generated/version.h
-- Zephyr version: 3.5.99 (/home/ycsin/zephyrproject/zephyr), build: zephyr-v3.5.0-1517-g5e77c952298b
[110/111] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
             ROM:       50456 B       256 KB     19.25%
             RAM:       27424 B       256 KB     10.46%
        IDT_LIST:          0 GB         2 KB      0.00%
[110/111] cd /home/ycsin/zephyrproject/build && /usr/bin/renode --disa...ect/zephyr/boards/riscv/renode_riscv32/support/renode_riscv32.resc; s'
15:49:18.8777 [INFO] Loaded monitor commands from: /opt/renode/scripts/monitor.py
15:49:18.9818 [INFO] Including script: /home/ycsin/zephyrproject/zephyr/boards/riscv/renode_riscv32/support/renode_riscv32.resc
15:49:19.0161 [INFO] System bus created.
15:49:20.5801 [WARNING] Undefined instruction set: _.
15:49:20.5802 [WARNING] Undefined instruction set: Z.
15:49:20.5804 [WARNING] Undefined instruction set: R.
15:49:20.5805 [WARNING] Undefined instruction set: _.
15:49:20.5807 [WARNING] Undefined instruction set: Z.
15:49:20.5808 [WARNING] Undefined instruction set: E.
15:49:20.5809 [WARNING] Undefined instruction set: N.
15:49:20.5811 [WARNING] Undefined instruction set: E.
15:49:21.9392 [INFO] sysbus: Loading segment of 33908 bytes length at 0x80000000.
15:49:21.9629 [INFO] sysbus: Loading segment of 16544 bytes length at 0x80008474.
15:49:21.9634 [INFO] sysbus: Loading segment of 4 bytes length at 0x8000C514.
15:49:21.9637 [INFO] sysbus: Loading segment of 10880 bytes length at 0x80040000.
15:49:22.0145 [INFO] cpu: Setting PC value to 0x80000000.
15:49:22.1383 [INFO] Renode-RISCV32: Machine started.
15:49:22.3817 [WARNING] plic0: Unhandled write to offset 0x200000, value 0x0.
15:49:22.4310 [WARNING] plic1: Unhandled write to offset 0x200000, value 0x0.
15:49:22.4802 [INFO] uart0: [host: 1.77s (+1.77s)|virt: 32.6ms (+32.6ms)] *** Booting Zephyr OS build zephyr-v3.5.0-1517-g5e77c952298b ***
15:49:22.4869 [INFO] uart0: [host: 1.78s (+7.09ms)|virt:  33.9ms (+1.3ms)] Running TESTSUITE interrupt_feature
15:49:22.4972 [INFO] uart0: [host: 1.79s (+10.35ms)|virt:  35.6ms (+1.7ms)] ===================================================================
15:49:22.5013 [INFO] uart0: [host:  1.79s (+3.86ms)|virt:  36.3ms (+0.7ms)] START - test_isr_dynamic
15:49:22.5117 [INFO] uart0: [host:  1.8s (+10.67ms)|virt:  37.4ms (+1.1ms)] installing dynamic ISR for IRQ 0
15:49:22.5252 [INFO] uart0: [host: 1.82s (+13.44ms)|virt:  38.7ms (+1.3ms)]  PASS - test_isr_dynamic in 0.002 seconds
15:49:22.5476 [INFO] uart0: [host: 1.84s (+22.47ms)|virt:  40.5ms (+1.8ms)] ===================================================================
15:49:22.5551 [INFO] uart0: [host:  1.85s (+7.19ms)|virt:  41.3ms (+0.8ms)] START - test_isr_offload_job
15:49:22.5629 [INFO] uart0: [host:   1.85s (+7.9ms)|virt:  42.9ms (+1.6ms)]  SKIP - test_isr_offload_job in 0.001 seconds
15:49:22.5765 [INFO] uart0: [host: 1.87s (+13.72ms)|virt:  44.6ms (+1.7ms)] ===================================================================
15:49:22.5807 [INFO] uart0: [host:  1.87s (+3.84ms)|virt:  45.5ms (+0.9ms)] START - test_isr_offload_job_identi
15:49:22.6642 [INFO] uart0: [host: 1.95s (+83.93ms)|virt: 63.4ms (+17.9ms)]  PASS - test_isr_offload_job_identi in 0.017 seconds
15:49:22.6791 [INFO] uart0: [host: 1.97s (+14.85ms)|virt:  65.1ms (+1.7ms)] ===================================================================
15:49:22.6829 [INFO] uart0: [host:  1.97s (+3.83ms)|virt:    66.1ms (+1ms)] START - test_isr_offload_job_multiple
15:49:22.7690 [INFO] uart0: [host: 2.06s (+85.73ms)|virt: 84.7ms (+18.6ms)]  PASS - test_isr_offload_job_multiple in 0.018 seconds
15:49:22.7754 [INFO] uart0: [host:  2.07s (+6.59ms)|virt:  86.3ms (+1.6ms)] ===================================================================
15:49:22.7818 [INFO] uart0: [host:  2.07s (+6.63ms)|virt:    87ms (+0.7ms)] START - test_nested_isr
15:49:22.7882 [INFO] uart0: [host:  2.08s (+6.39ms)|virt:  88.5ms (+1.5ms)]  SKIP - test_nested_isr in 0.001 seconds
15:49:22.7982 [INFO] uart0: [host:  2.09s (+9.97ms)|virt:  90.1ms (+1.6ms)] ===================================================================
15:49:22.8012 [INFO] uart0: [host:  2.09s (+2.96ms)|virt:    91.1ms (+1ms)] START - test_prevent_interruption
15:49:22.8128 [INFO] uart0: [host:   2.1s (+5.31ms)|virt:  91.8ms (+0.7ms)] locking interrupts
15:49:38.8663 [INFO] uart0: [host: 18.16s (+16.06s)|virt:   1.37s (+1.28s)] unlocking interrupts
15:49:38.8682 [INFO] uart0: [host: 18.16s (+1.94ms)|virt:   1.37s (+0.4ms)] timer fired
15:49:54.4216 [INFO] uart0: [host: 33.71s (+15.55s)|virt:   2.65s (+1.28s)]  PASS - test_prevent_interruption in 2.563 seconds
15:49:54.4294 [INFO] uart0: [host: 33.72s (+7.76ms)|virt:   2.66s (+1.6ms)] ===================================================================
15:49:54.4336 [INFO] uart0: [host: 33.72s (+4.17ms)|virt:   2.66s (+1.1ms)] START - test_sw_isr_irq_parent_table_dev
15:49:54.5257 [INFO] uart0: [host: 33.82s (+92.13ms)|virt:  2.69s (+30.9ms)]  PASS - test_sw_isr_irq_parent_table_dev in 0.030 seconds
15:49:54.5317 [INFO] uart0: [host:  33.82s (+6.03ms)|virt:   2.69s (+1.6ms)] ===================================================================
15:49:54.5390 [INFO] uart0: [host:  33.83s (+7.11ms)|virt:   2.69s (+1.1ms)] START - test_sw_isr_irq_parent_table_idx
15:49:54.6410 [INFO] uart0: [host:    33.93s (+0.1s)|virt:    2.73s (+37ms)]  PASS - test_sw_isr_irq_parent_table_idx in 0.036 seconds
15:49:54.6476 [INFO] uart0: [host:  33.94s (+6.71ms)|virt:   2.73s (+1.6ms)] ===================================================================
15:49:54.6535 [INFO] uart0: [host:  33.94s (+5.83ms)|virt:     2.73s (+1ms)] TESTSUITE interrupt_feature succeeded
15:49:54.6548 [INFO] uart0: [host:  33.95s (+1.43ms)|virt:   2.73s (+0.4ms)] 
15:49:54.6595 [INFO] uart0: [host:  33.95s (+4.61ms)|virt:   2.73s (+0.9ms)] ------ TESTSUITE SUMMARY START ------
15:49:54.6598 [INFO] uart0: [host:  33.95s (+0.35ms)|virt:   2.73s (+0.1ms)] 
15:49:54.6772 [INFO] uart0: [host: 33.97s (+17.45ms)|virt:   2.74s (+3.2ms)] SUITE PASS - 100.00% [interrupt_feature]: pass = 6, fail = 0, skip = 2, total = 8 duration = 2.668 seconds
15:49:54.6858 [INFO] uart0: [host:  33.98s (+8.42ms)|virt:   2.74s (+1.9ms)]  - PASS - [interrupt_feature.test_isr_dynamic] duration = 0.002 seconds
15:49:54.6971 [INFO] uart0: [host: 33.99s (+11.45ms)|virt:     2.74s (+2ms)]  - SKIP - [interrupt_feature.test_isr_offload_job] duration = 0.001 seconds
15:49:54.7078 [INFO] uart0: [host:    34s (+10.62ms)|virt:   2.74s (+2.2ms)]  - PASS - [interrupt_feature.test_isr_offload_job_identi] duration = 0.017 seconds
15:49:54.7210 [INFO] uart0: [host: 34.01s (+13.27ms)|virt:   2.74s (+2.3ms)]  - PASS - [interrupt_feature.test_isr_offload_job_multiple] duration = 0.018 seconds
15:49:54.7306 [INFO] uart0: [host:  34.02s (+9.42ms)|virt:   2.75s (+1.9ms)]  - SKIP - [interrupt_feature.test_nested_isr] duration = 0.001 seconds
15:49:54.7439 [INFO] uart0: [host: 34.03s (+13.38ms)|virt:   2.75s (+2.1ms)]  - PASS - [interrupt_feature.test_prevent_interruption] duration = 2.563 seconds
15:49:54.7546 [INFO] uart0: [host: 34.05s (+10.76ms)|virt:   2.75s (+2.3ms)]  - PASS - [interrupt_feature.test_sw_isr_irq_parent_table_dev] duration = 0.030 seconds
15:49:54.7669 [INFO] uart0: [host: 34.06s (+12.35ms)|virt:   2.75s (+2.4ms)]  - PASS - [interrupt_feature.test_sw_isr_irq_parent_table_idx] duration = 0.036 seconds
15:49:54.7671 [INFO] uart0: [host:  34.06s (+0.24ms)|virt:      2.75s (+0s)] 
15:49:54.7709 [INFO] uart0: [host:  34.06s (+3.65ms)|virt:   2.75s (+0.9ms)] ------ TESTSUITE SUMMARY END ------
15:49:54.7712 [INFO] uart0: [host:  34.06s (+0.13ms)|virt:      2.75s (+0s)] 
15:49:54.7811 [INFO] uart0: [host: 34.07s (+10.15ms)|virt:   2.75s (+1.7ms)] ===================================================================
15:49:54.7858 [INFO] uart0: [host:  34.08s (+4.64ms)|virt:   2.76s (+0.7ms)] PROJECT EXECUTION SUCCESSFUL
15:51:41.2940 [INFO] Renode-RISCV32: Machine paused.
15:51:41.3263 [INFO] Renode-RISCV32: Disposed.
```

### CODE REVIEW

**Review by jgl-meta — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Comment by ycsin** [comment]
Force pushed to fix compilation error for xtensa's `CONFIG_NUM_IRQS` implementation

**Review by cfriedt — APPROVED** [review]
LGTM

**Comment by ycsin** [comment]
@fkokosinski could you please take a look?

**Review by fkokosinski — APPROVED** [review]
From the perspective of RISC-V in Zephyr - LGTM. I suggest that maintainers of other archs in Zephyr should also have a chance to look at these changes, though.

**Comment by cfriedt** [comment]
@nashif, @dcpleung - can you please review? IIRC, it's mainly Xtensa / Intel ADSP that this affects other than RISC-V, and those are areas you maintain.

**Comment by nashif** [comment]
> @nashif, @dcpleung - can you please review? IIRC, it's mainly Xtensa / Intel ADSP that this affects other than RISC-V, and those are areas you maintain.

will do

**Review by dcpleung — COMMENTED** [review]
LGTM. But will need someone from the Intel audio team to test. @kv2019i ?

**Comment by cfriedt** [comment]
Adding `DNM` til EOD to wait for @kv2019i to test

**Review by jgl-meta — APPROVED** [review]

**Review by nashif — APPROVED** [review]
lgtm, but lets wait for some hw testing.

**Review by kv2019i — APPROVED** [review]
Looks good, test results OK on SOF xtensa platforms.

**Comment by ycsin** [comment]
Removing `DNM`

**Comment by henrikbrixandersen** [comment]
@andyross Please review.

File: arch/common/multilevel_irq.c [hunk 1]
```diff
@@ -6,35 +6,41 @@
...
 
 #include <zephyr/device.h>
+#include <zephyr/irq.h>
 #include <zephyr/sw_isr_table.h>
+#include <zephyr/sys/__assert.h>
 #include <zephyr/sys/util.h>
 
+BUILD_ASSERT((CONFIG_NUM_2ND_LEVEL_AGGREGATORS * CONFIG_MAX_IRQ_PER_AGGREGATOR) <=
+		     BIT(CONFIG_2ND_LEVEL_INTERRUPT_BITS),
+	     "L2 bits not enough to cover the number of L2 IRQs");
```
Comment (cfriedt) [inline arch/common/multilevel_irq.c:16]: I think both of these should be in units of "per aggregator". I.e. that it's incorrect to multiply the LHS side of this relationship with `CONFIG_NUM_2ND_LEVEL_AGGREGATORS`.
```suggestion
BUILD_ASSERT(CONFIG_MAX_IRQ_PER_AGGREGATOR <= BIT(CONFIG_2ND_LEVEL_INTERRUPT_BITS),
	     "L2 bits not enough to cover the number of L2 IRQs");
```

File: arch/common/multilevel_irq.c [hunk 2]
```diff
@@ -54,12 +60,16 @@ const struct _irq_parent_entry _lvl2_irq_list[CONFIG_NUM_2ND_LEVEL_AGGREGATORS]
 	= { LISTIFY(CONFIG_NUM_2ND_LEVEL_AGGREGATORS, CAT_2ND_LVL_LIST, (,),
 		CONFIG_2ND_LVL_ISR_TBL_OFFSET) };
 
+#ifdef CONFIG_3RD_LEVEL_INTERRUPTS
+
+BUILD_ASSERT((CONFIG_NUM_3RD_LEVEL_AGGREGATORS * CONFIG_MAX_IRQ_PER_AGGREGATOR) <=
+		     BIT(CONFIG_3RD_LEVEL_INTERRUPT_BITS),
+	     "L3 bits not enough to cover the number of L3 IRQs");
```
Comment (cfriedt) [inline arch/common/multilevel_irq.c:67]: I think both of these should be in units of "per aggregator". I.e. that it's incorrect to multiply the LHS side of this relationship with `CONFIG_NUM_2ND_LEVEL_AGGREGATORS`.
```suggestion
#ifdef CONFIG_3RD_LEVEL_INTERRUPTS

BUILD_ASSERT(CONFIG_MAX_IRQ_PER_AGGREGATOR <= BIT(CONFIG_3RD_LEVEL_INTERRUPT_BITS),
	     "L3 bits not enough to cover the number of L3 IRQs");
```

File: arch/common/multilevel_irq.c (hunk 1 above)
Comment (ycsin) [reply arch/common/multilevel_irq.c:16]: That's how the offset is calculated for each aggregator in the multi-level LUT:

https://github.com/zephyrproject-rtos/zephyr/blob/9102821fbe282d1b994ee21c0e3015c885de3d58/arch/common/multilevel_irq.c#L48-L55

`CONFIG_MAX_IRQ_PER_AGGREGATOR` basically means "Max IRQs per PLIC" in the context of RISCV PLIC, to get the total number of IRQs supported in the 2nd level in software, we need to multiply `CONFIG_MAX_IRQ_PER_AGGREGATOR` by the number of 2nd level aggregator (`CONFIG_NUM_2ND_LEVEL_AGGREGATORS`)

If we have 1 PLIC that supports 1024 IRQs, we will have 1024 IRQs in the 2nd level, which need 10 bits in the 2nd level (2<sup>10</sup> = 1024) to encode.

If we have another instance of such PLIC, we will have 2048 IRQs in the 2nd level, requiring 11 bits in the 2nd level to encode (2<sup>11</sup> = 2048).

Comment (cfriedt) [reply arch/common/multilevel_irq.c:16]: Hmm... It's a bit counter intuitive.

Fine for now I guess but when we start looking at chained interrupts, we should probably scrap this

**Review by andyross — APPROVED** [review]
This all makes sense, though I'll be honest that I still find this interface extremely confusing and think what it really needs is a rewrite.

**Comment by ycsin** [comment]
Quoting @andyross [comment](https://discord.com/channels/720317445772017664/997527108844798012/1182678097938620416) in Discord for safekeeping:

> Thanks for the ping, +1.  Though while you're in there if you want to give some thought to a rework, I wouldn't complain.  Basically the tree of controllers has all the info it needs in DTS already, all those kconfigs are duplicated and fragile.  Ideally you'd just set it up in DTS and then Zephyr would (1) provide an API to convert between "Nth interrupt on this controller" and "global IRQ number" spaces and (2) emit all the routine code automatically so the SOC layer doesn't have to do it.

**Pull request merged**